src/form.js

```diff
@@ -1022,35 +1037,39 @@ define([
             }
             var _this = this,
                 seen = {},
-                ufids = {},
-                undoUfids = {};
+                ufids = {};
             this.undomanager.resetUndo();
-            _.each(mugs, function (mug) {
-                _this._addToUndoManager(mug, undoUfids);
-            });
+            _this._addToUndoManager(mugs);
             _.each(mugs, function (mug) {
                 _this._removeMugFromForm(mug, ufids, false);
             });
             this._logicManager.forEachReferencingProperty(ufids, breakReferences);
         },
-        _addToUndoManager: function(mug, ufids) {
-            if (ufids.hasOwnProperty(mug.ufid)) {
-                return; // already removed
-            }
-            ufids[mug.ufid] = null;
-            var node = this.tree.getNodeFromMug(mug),
-                parentMug = mug.parentMug,
-                hasChildren = false,
-                previousSibling = mug.previousSibling,
-                position = previousSibling === parentMug ? 'first' : 'after';
-            this.undomanager.appendMug(mug, previousSibling, position);
-            if (node) {
-                var children = node.getChildrenMugs();
-                hasChildren = children.length > 0;
-                for (var i = 0; i < children.length; i++) {
-                    this._addToUndoManager(children[i], ufids);
+        _addToUndoManager: function(mugs) {
+            function pushUndoItem(mug) {
+                if (ufids.hasOwnProperty(mug.ufid)) {
+                    return; // already removed
+                }
+                ufids[mug.ufid] = null;
+                var node = tree.getNodeFromMug(mug),
+                    parentMug = mug.parentMug,
+                    hasChildren = false,
+                    previousSibling = mug.previousSibling,
+                    position = previousSibling === parentMug ? 'first' : 'after';
+                stack.push([mug, previousSibling, position]);
+                if (node) {
+                    var children = node.getChildrenMugs();
+                    hasChildren = children.length > 0;
+                    for (var i = 0; i < children.length; i++) {
+                        pushUndoItem(children[i], ufids);
+                    }
```

Comment:
This didn't change much. Best [view with `w=1`](https://github.com/dimagi/Vellum/pull/857/commits/010cada40cc75fde9d093dd1087dc526f685a968?w=1)